kv/kvserver: TestStoreRangeWaitForApplication failed

kv/kvserver.TestStoreRangeWaitForApplication [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/18220869?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/18220869?buildTab=artifacts#/) on master @ [e30d684e1625734f2ad9e00bad375e70e6a317db](https://github.com/cockroachdb/cockroach/commits/e30d684e1625734f2ad9e00bad375e70e6a317db):

Fatal error:

```
panic: test timed out after 15m0s
running tests:
	TestStoreRangeWaitForApplication (14m51s)
	TestStoreRangeWaitForApplication/lease-type=LeaseLeader (14m49s)

```
Stack: 

```
goroutine 120450 [running]:
testing.(*M).startAlarm.func1()
	GOROOT/src/testing/testing.go:2366 +0x30c
created by time.goFunc
	GOROOT/src/time/sleep.go:177 +0x38
```
<details><summary>Log preceding fatal error</summary>
<p>

```
=== RUN   TestStoreRangeWaitForApplication
    test_log_scope.go:165: test logs captured to: /artifacts/tmp/_tmp/ea6fe1b0541b7dd6caeb699ed27d18f6/logTestStoreRangeWaitForApplication2064169084
    test_log_scope.go:76: use -show-logs to present logs inline
=== RUN   TestStoreRangeWaitForApplication/lease-type=LeaseEpoch
=== RUN   TestStoreRangeWaitForApplication/lease-type=LeaseExpiration
=== RUN   TestStoreRangeWaitForApplication/lease-type=LeaseLeader
    testcluster.go:1178: lease is now of type: LeaseLeader
```
</p>
</details>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
/cc @cockroachdb/kv-triage
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestStoreRangeWaitForApplication.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)

</sub>


Jira issue: CRDB-45678

kv/kvserver.TestStoreRangeWaitForApplication [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/18274205?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/18274205?buildTab=artifacts#/) on master @ [bea204ff6caead5a3b47db83d68f2e6e0231b84d](https://github.com/cockroachdb/cockroach/commits/bea204ff6caead5a3b47db83d68f2e6e0231b84d):

Fatal error:

```
panic: test timed out after 15m0s
running tests:
	TestStoreRangeWaitForApplication (14m48s)
	TestStoreRangeWaitForApplication/lease-type=LeaseLeader (14m46s)

```
Stack: 

```
goroutine 107601 [running]:
testing.(*M).startAlarm.func1()
	GOROOT/src/testing/testing.go:2366 +0x30c
created by time.goFunc
	GOROOT/src/time/sleep.go:177 +0x38
```
<details><summary>Log preceding fatal error</summary>
<p>

```
=== RUN   TestStoreRangeWaitForApplication
    test_log_scope.go:165: test logs captured to: /artifacts/tmp/_tmp/ea6fe1b0541b7dd6caeb699ed27d18f6/logTestStoreRangeWaitForApplication297844177
    test_log_scope.go:76: use -show-logs to present logs inline
=== RUN   TestStoreRangeWaitForApplication/lease-type=LeaseEpoch
=== RUN   TestStoreRangeWaitForApplication/lease-type=LeaseExpiration
=== RUN   TestStoreRangeWaitForApplication/lease-type=LeaseLeader
    testcluster.go:1178: lease is now of type: LeaseLeader
```
</p>
</details>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestStoreRangeWaitForApplication.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)

</sub>

--- a/pkg/kv/kvserver/client_raft_test.go
+++ b/pkg/kv/kvserver/client_raft_test.go
@@ -4596,12 +4596,6 @@ func TestStoreRangeWaitForApplication(t *testing.T) {
 
 	testutils.RunValues(t, "lease-type", roachpb.TestingAllLeaseTypes(),
 		func(t *testing.T, leaseType roachpb.LeaseType) {
-			if leaseType == roachpb.LeaseLeader {
-				// TODO(ibrahim): Find out why this test is very slow with leader leases
-				// if run under race/deadlock.
-				skip.UnderRace(t, "slow with leader leases")
-				skip.UnderDeadlock(t, "slow with leader leases")
-			}
 			var filterRangeIDAtomic int64
 
 			ctx := context.Background()
@@ -4629,8 +4623,14 @@ func TestStoreRangeWaitForApplication(t *testing.T) {
 						Settings: settings,
 						Knobs: base.TestingKnobs{
 							Store: &kvserver.StoreTestingKnobs{
+								TestingRequestFilter: testingRequestFilter,
+								// This test relies on a stable LAI, so we need to disable GC
+								// queue and async intent resolution to avoid the LAI being
+								// incremented in the middle of the test.
 								DisableReplicaGCQueue: true,
-								TestingRequestFilter:  testingRequestFilter,
+								IntentResolverKnobs: kvserverbase.IntentResolverTestingKnobs{
+									DisableAsyncIntentResolution: true,
+								},
 							},
 						},
 					},
@@ -4649,16 +4649,10 @@ func TestStoreRangeWaitForApplication(t *testing.T) {
 				t.Fatal(err)
 			}
 
+			// We wait for lease upgrade to avoid the lease being upgraded after we
+			// capture the LAI.
+			tc.MaybeWaitForLeaseUpgrade(ctx, t, desc)
 			leaseIndex0 := repl0.GetLeaseAppliedIndex()
-			if leaseType == roachpb.LeaseLeader {
-				// When running with leader leases, when there is no leader, we acquire
-				// an expiration based lease, and then acquiring a leader lease.
-				// Therefore, the lease applied index is incremented by an extra 1.
-				leaseIndex0++
-				// Before blocking requests, make sure that the expiration lease gets
-				// upgraded to a leader lease.
-				tc.WaitForLeaseUpgrade(ctx, t, desc)
-			}
 
 			atomic.StoreInt64(&filterRangeIDAtomic, int64(desc.RangeID))
 			type target struct {

